### ISSUE TITLE
🐛 Bug Report: UrlReader service is not working out-of-the-box with "factories" dependency

### ISSUE THREAD

**Issue by humbleice** [backstage/backstage#28177]
### 📜 Description

When using the new backend system, plugins that use the UrlReaderService don't work out of the box.  This was found when migrating from the legacy backend to the new backend.

### 👍 Expected behavior

When starting the backend, there should not be an error when loading plugins that rely on the out-of-box UrlReaderService.

### 👎 Actual Behavior with Screenshots

Using out-of-the-box plugins which rely on the UrlReaderService, the following error message is provided:
> Error: Failed to instantiate service 'core.urlReader' for 'catalog' because the following dependent services are missing: 'core.urlReader.factories'

![image](https://github.com/user-attachments/assets/b8291098-af35-423a-9638-961b91dcf92e)

### 👟 Reproduction steps

Presumably, a new out-of-the-box project using the new backend along with plugins that depend on the urlReader should reproduce.
e.g.,
```
const backend = createBackend();
backend.add(import('@backstage/plugin-scaffolder-backend'));
backend.add(import('@backstage/plugin-techdocs-backend'));
backend.start()
```
then in terminal
```
yarn start-backend
```

### 📃 Provide the context for the Bug.

I came across this bug while upgrading from the legacy backend to the new backend.  The plugins were working fine before the migration, but after changing out-of-the-box plugins to use the new `backend.add(import('@backstage/some-plugin'));` pattern, the backend started to throw these errors.

### Debugging Process

Since the error was related to "core.urlReader", I found the [documentation](https://backstage.io/docs/backend-system/core-services/url-reader/) about this service to see if there were any additional steps I needed to make for the new backend.  There wasn't anything required steps mentioned, but there is an example of how to create custom UrlReader, which I eventually use as a workaround for this bug.

With more digging around, I determined the "core.urlReader.factories" dependency from the error is actually related to a relatively new service reference called [urlReaderFactoriesServiceRef](https://github.com/backstage/backstage/blob/master/packages/backend-defaults/src/entrypoints/urlReader/urlReaderServiceFactory.ts#L42).  It seems this service is [not created automatically](https://github.com/backstage/backstage/blob/master/packages/backend-defaults/src/CreateBackend.ts#L54) when starting the new backend, which appears to be why the dependency is missing.

To fix the missing dependency issue, I decided to use the example from the aforementioned UrlReaderService documentation, which adds the service explicitly to the backend.  Following the documentation helped me move past the missing dependency error, but a new error was occurring:

> Failed to instantiate service 'core.urlReader' for 'azure-devops' because the factory function threw an error, TypeError: factories.concat is not a function

With more digging around, I concluded this issue was likely related to [the code that combines the custom UrlReader factories from the default UrlReader factories](https://github.com/backstage/backstage/blob/master/packages/backend-defaults/src/entrypoints/urlReader/lib/UrlReaders.ts#L91).  It seems the factory instance from my `urlReaderFactoriesServiceRef` service factory was providing only the factory object and not an array of objects, even though an array is expected.

From this point, I needed to deviate from the documentation and out-of-the-box code to come up with a workaround.  To fix this, I created a new ServiceRef that's similar to `urlReaderFactoriesServiceRef` but is defined like so:
```
createServiceRef<ReaderFactory[]>({
  id: "core.urlReader.factories",
  scope: "plugin",
  multiton: true
});
```
The only change here is using `ReaderFactory[]` instead of `ReaderFactory`.

Additionally, I needed to update my customServiceFactory to return `[CustomUrlReader.factory]` instead of `CustomUrlReader.factory`.

With these changes the issue was fixed, and the backend could start without failure.

### Full Code Workaround
This is added in `packages/backend/src/index.ts` after creating the backend. (There's probably a more succinct workaround but this did the job for me).
```
let urlReaderFactoriesServiceRef = createServiceRef<ReaderFactory[]>({
  id: "core.urlReader.factories",
  scope: "plugin",
  multiton: true
});

class CustomUrlReader implements UrlReaderService {
  static factory: ReaderFactory = ({ config, treeResponseFactory }) => {
    const reader = new CustomUrlReader(config);
    const predicate = (url: URL) => url.host === 'myCustomDomain';
    return [{ reader, predicate }];
  };

  constructor(private readonly config: Config) {}
    readUrl(url: string, options?: UrlReaderServiceReadUrlOptions): Promise<UrlReaderServiceReadUrlResponse> {
        throw new Error('Method not implemented.');
    }
    readTree(url: string, options?: UrlReaderServiceReadTreeOptions): Promise<UrlReaderServiceReadTreeResponse> {
        throw new Error('Method not implemented.');
    }
    search(url: string, options?: UrlReaderServiceSearchOptions): Promise<UrlReaderServiceSearchResponse> {
        throw new Error('Method not implemented.');
    }
}

const customReader = createServiceFactory({
  service: urlReaderFactoriesServiceRef,
  deps: {},
  async factory() {
    return [CustomUrlReader.factory];
  },
});
backend.add(customReader);
```

### Diagnosis

1) UrlReaderService, which is automatically loaded by backend, depends on `core.urlReader.factories` which is not automatically loaded by backend.  This causes problems for plugins trying to use urlReader assuming an out-of-the-box setup.

2) urlReaderFactoriesServiceRef defines a type ReaderFactory, when it likely should define ReaderFactory[] since an array is expected in places where it's used in the backstage code.

### 🖥️ Your Environment

```
OS:   Linux 5.15.167.4-microsoft-standard-WSL2 - linux/x64
node: v20.17.0
yarn: 4.5.1
cli:  0.29.3 (installed)
backstage:  1.33.0

Dependencies:
  @backstage/app-defaults                                            1.5.13
  @backstage/backend-app-api                                         0.8.0
  @backstage/backend-common                                          0.21.7, 0.22.0, 0.23.3, 0.24.1, 0.25.0
  @backstage/backend-defaults                                        0.2.18, 0.5.3
  @backstage/backend-dev-utils                                       0.1.5
  @backstage/backend-openapi-utils                                   0.3.0
  @backstage/backend-plugin-api                                      0.6.21, 0.7.0, 0.8.1, 1.0.2
  @backstage/backend-tasks                                           0.5.27, 0.6.1
  @backstage/backend-test-utils                                      1.1.0
  @backstage/catalog-client                                          1.8.0
  @backstage/catalog-model                                           1.7.1
  @backstage/cli-common                                              0.1.15
  @backstage/cli-node                                                0.2.10
  @backstage/cli                                                     0.29.0, 0.29.3
  @backstage/config-loader                                           1.9.2
  @backstage/config                                                  1.3.0
  @backstage/core-app-api                                            1.15.2
  @backstage/core-compat-api                                         0.3.2, 0.3.3
  @backstage/core-components                                         0.12.5, 0.14.10, 0.15.1, 0.16.0, 0.16.1
  @backstage/core-plugin-api                                         1.10.1
  @backstage/e2e-test-utils                                          0.1.1
  @backstage/errors                                                  1.2.5
  @backstage/eslint-plugin                                           0.1.10
  @backstage/frontend-app-api                                        0.10.1, 0.10.2
  @backstage/frontend-defaults                                       0.1.2, 0.1.3
  @backstage/frontend-plugin-api                                     0.9.1, 0.9.2
  @backstage/frontend-test-utils                                     0.2.2, 0.2.3
  @backstage/integration-aws-node                                    0.1.13
  @backstage/integration-react                                       1.2.1
  @backstage/integration                                             1.15.2
  @backstage/plugin-api-docs                                         0.12.0
  @backstage/plugin-app-backend                                      0.4.0
  @backstage/plugin-app-node                                         0.1.27
  @backstage/plugin-app                                              0.1.2, 0.1.3
  @backstage/plugin-auth-backend-module-atlassian-provider           0.3.2
  @backstage/plugin-auth-backend-module-auth0-provider               0.1.2
  @backstage/plugin-auth-backend-module-aws-alb-provider             0.3.0
  @backstage/plugin-auth-backend-module-azure-easyauth-provider      0.2.2
  @backstage/plugin-auth-backend-module-bitbucket-provider           0.2.2
  @backstage/plugin-auth-backend-module-bitbucket-server-provider    0.1.2
  @backstage/plugin-auth-backend-module-cloudflare-access-provider   0.3.2
  @backstage/plugin-auth-backend-module-gcp-iap-provider             0.3.2
  @backstage/plugin-auth-backend-module-github-provider              0.2.2
  @backstage/plugin-auth-backend-module-gitlab-provider              0.2.2
  @backstage/plugin-auth-backend-module-google-provider              0.2.2
  @backstage/plugin-auth-backend-module-guest-provider               0.2.2
  @backstage/plugin-auth-backend-module-microsoft-provider           0.2.2
  @backstage/plugin-auth-backend-module-oauth2-provider              0.3.2
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider        0.2.2
  @backstage/plugin-auth-backend-module-oidc-provider                0.3.2
  @backstage/plugin-auth-backend-module-okta-provider                0.1.2
  @backstage/plugin-auth-backend-module-onelogin-provider            0.2.2
  @backstage/plugin-auth-backend                                     0.24.0
  @backstage/plugin-auth-node                                        0.4.17, 0.5.4
  @backstage/plugin-auth-react                                       0.1.8
  @backstage/plugin-bitbucket-cloud-common                           0.2.25
  @backstage/plugin-catalog-backend-module-github                    0.7.7
  @backstage/plugin-catalog-backend-module-gitlab                    0.5.0
  @backstage/plugin-catalog-backend-module-msgraph                   0.6.4
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model   0.2.2
  @backstage/plugin-catalog-backend                                  1.28.0
  @backstage/plugin-catalog-common                                   1.1.1
  @backstage/plugin-catalog-graph                                    0.4.12
  @backstage/plugin-catalog-import                                   0.12.6
  @backstage/plugin-catalog-node                                     1.14.0
  @backstage/plugin-catalog-react                                    1.14.1, 1.14.2
  @backstage/plugin-catalog                                          1.25.0
  @backstage/plugin-devtools-backend                                 0.4.2
  @backstage/plugin-devtools-common                                  0.1.13
  @backstage/plugin-events-backend-module-github                     0.2.14
  @backstage/plugin-events-backend-module-gitlab                     0.2.14
  @backstage/plugin-events-backend                                   0.3.16
  @backstage/plugin-events-node                                      0.3.10, 0.4.5
  @backstage/plugin-home-react                                       0.1.19
  @backstage/plugin-kubernetes-common                                0.7.6, 0.8.3
  @backstage/plugin-kubernetes-react                                 0.4.4
  @backstage/plugin-kubernetes                                       0.11.16
  @backstage/plugin-org                                              0.6.32
  @backstage/plugin-permission-backend                               0.5.51
  @backstage/plugin-permission-common                                0.7.14, 0.8.2
  @backstage/plugin-permission-node                                  0.8.5
  @backstage/plugin-permission-react                                 0.4.28
  @backstage/plugin-proxy-backend                                    0.5.8
  @backstage/plugin-scaffolder-backend-module-azure                  0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud        0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket-server       0.2.2
  @backstage/plugin-scaffolder-backend-module-bitbucket              0.3.2
  @backstage/plugin-scaffolder-backend-module-confluence-to-markdown 0.3.2
  @backstage/plugin-scaffolder-backend-module-gerrit                 0.2.2
  @backstage/plugin-scaffolder-backend-module-gitea                  0.2.2
  @backstage/plugin-scaffolder-backend-module-github                 0.5.2
  @backstage/plugin-scaffolder-backend-module-gitlab                 0.6.1
  @backstage/plugin-scaffolder-backend                               1.27.0
  @backstage/plugin-scaffolder-common                                1.5.7
  @backstage/plugin-scaffolder-node                                  0.4.11, 0.6.0
  @backstage/plugin-scaffolder-react                                 1.14.0
  @backstage/plugin-scaffolder                                       1.27.0
  @backstage/plugin-search-backend-module-catalog                    0.2.5
  @backstage/plugin-search-backend-module-pg                         0.5.38
  @backstage/plugin-search-backend-module-techdocs                   0.3.2
  @backstage/plugin-search-backend-node                              1.3.5
  @backstage/plugin-search-backend                                   1.7.0
  @backstage/plugin-search-common                                    1.2.15
  @backstage/plugin-search-react                                     1.8.2
  @backstage/plugin-search                                           1.4.19
  @backstage/plugin-signals-backend                                  0.2.3
  @backstage/plugin-signals-node                                     0.1.14
  @backstage/plugin-signals-react                                    0.0.7
  @backstage/plugin-techdocs-backend                                 1.11.2
  @backstage/plugin-techdocs-common                                  0.1.0
  @backstage/plugin-techdocs-module-addons-contrib                   1.1.17
  @backstage/plugin-techdocs-node                                    1.12.13
  @backstage/plugin-techdocs-react                                   1.2.10
  @backstage/plugin-techdocs                                         1.11.1
  @backstage/plugin-user-settings-common                             0.0.1
  @backstage/plugin-user-settings                                    0.8.15
  @backstage/release-manifests                                       0.0.11
  @backstage/test-utils                                              1.7.1, 1.7.2
  @backstage/theme                                                   0.2.19, 0.5.7, 0.6.1, 0.6.2
  @backstage/types                                                   1.2.0
  @backstage/version-bridge                                          1.0.10
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Hi @humbleice, thanks for the detailed issue! 

Based on what your are saying we should be seeing this issue with a fresh Backstage instance created with `npx @backstage/create-app@latest` but I'm not as I just gave it a try. We also have the same setup for the Demo site which is also working: https://demo.backstage.io/ (code is [here](https://github.com/backstage/demo/blob/master/packages/backend/src/index.ts)).

I am wondering if all the duplicate packages are part of the issue, this is generally concerning:

```text
@backstage/backend-common                                          0.21.7, 0.22.0, 0.23.3, 0.24.1, 0.25.0
@backstage/backend-defaults                                        0.2.18, 0.5.3
@backstage/backend-plugin-api                                      0.6.21, 0.7.0, 0.8.1, 1.0.2
@backstage/backend-tasks                                           0.5.27, 0.6.1
```

I would try running `yarn dedupe` and see if that gets a better result 🤔

**Comment by humbleice** [comment]
@awanlin Thanks for verifying this isn't an issue with a fresh instance.  Based on the findings from my investigation, it looked like it might be.  I'll see about deduping my packages and see if that helps.  Otherwise, I'll see if I can figure out good reproduction steps for this ticket.

**Comment by humbleice** [comment]
@awanlin, I was able to resolve this.  After doing `yarn backstage-cli version:bump` and `yarn dedupe` the problem persisted.  Then I noticed the following in my main packages.json file:
```
  "resolutions": {
    ...
    "@backstage/backend-app-api": "^0.8.0",
    "@backstage/plugin-catalog-backend": "^1.26.0",
    ...
  },
```
I'm not sure why these were added, but removing them and reinstalling fixed the issue with the UrlReader, and now the OOB implementation works as expected.

This could probably be closed as a non-issue.  Thanks!
